src/core/lwt.ml

```diff
@@ -22,1307 +22,2746 @@
+   the inner and outer ['b t]s behave identically.
+
+   This is accomplished by "unifying" them, which simply means: making one of
+   the promises point to the other. One of the promises thus becomes a "proxy,"
+   and the other is its "underlying" promise.
+
+   After that, all operations that would be performed by Lwt on the proxy are
+   instead performed on the underlying promise. This is ensured by the numerous
+   calls to the internal function [underlying] in this file.
+
+   Because of the pervasive use of [underlying], proxies and unification can be
+   more or less ignored on a first reading the code. However, becoming a proxy
+   is a kind of state change, and any promise that is returned by a callback
+   to [bind], or to a similar Lwt function, might become a proxy. That means:
+   just about any promise that is handed to the user, might become a proxy
+   promise by the next time Lwt sees it. This is important for reasoning about
+   possible state changes in implementation of Lwt, and is referenced in some
+   implementation detail comments.
+
+
+   8. Sequence-associated storage
+
+   Lwt has a global key-value map. The map can be preserved across sequential
+   composition functions, so that it has the same state in the user's callback
+   [f] as it did at the time the user called [Lwt.bind p f].
+
+   The details are pretty straightforward, and discussed in module
+   [Sequence_associated_storage]. The main thing to be aware of is the many
+   references to [current_storage] throughout Lwt, which are needed to properly
+   save and restore the mapping.
+
+
+   9. Type system abuse
+
+   The implementation uses the type system somewhat extensively. For example,
+   the promise state is a GADT which encodes the state in its type parameters.
+   Thus, if you do [let p = underlying p], the shadowing reference [p] is
+   statically known *not* to be a proxy, and the compiler knows that the
+   corresponding match case [Unified_with _] is impossible.
+
+   The external promise type, ['a t], and the external resolver type, ['a u],
+   are not GADTs. Furthermore, they are, respectively, covariant and
+   contravariant in ['a], while the internal promise type is invariant in
+   ['a]. For these reasons, there are nasty casts between ['a t], ['a u], and
+   the internal promise type. The implementation is, of course, written in
+   terms of the internal type.
+
+   Casting from an ['a t] to an internal promise produces a reference for
+   which the state is "unknown": this is simulated with a helper GADT, which
+   encodes existential types. There are several similar casts, which are used
+   to document possible state changes between the time a promise is created,
+   and the later time it is used in a callback. You can see these casts in
+   action in [Lwt.bind]. The cast syntax is pretty light, and, besides being
+   commented in [bind], all such casts are documented in modules [Public_types]
+   and [Basic_helpers].
```

Comment:
I honestly don't know if they do, but I ran code with casts through some performance tests with Flambda, and everything seemed fine. We'll have to:

1. Keep track of Flambda development (I've already started keeping closer track of compiler development, because it would have saved me like a week of time on the Skylake bug, not to mention others involved in debugging that on and off for Lwt).
2. Run future stress tests (#372) with Flambda turned on as well.

Even though there is talk of breaking `Obj.magic` and the like by Flambda, I don't think it can actually do that, as enough code uses it. If it does, I'm guessing there will be flags and/or annotations to disable parts of Flambda, which would also give Lwt time to find better ways of expressing things.

Finally, I should also note that the casting here is worse, but not much worse, than in `lwt.ml` before the PR (which used `%identity` a lot). In the worst case, we can return to the previous level of casting by abandoning the GADT, or maybe doing something smarter with it.